telegram/_shared.py

```diff
@@ -37,40 +48,89 @@ class UsersShared(TelegramObject):
         request_id (:obj:`int`): Identifier of the request.
-        user_ids (Sequence[:obj:`int`]): Identifiers of the shared users. These numbers may have
-            more than 32 significant bits and some programming languages may have difficulty/silent
-            defects in interpreting them. But they have at most 52 significant bits, so 64-bit
-            integers or double-precision float types are safe for storing these identifiers. The
-            bot may not have access to the users and could be unable to use these identifiers,
-            unless the users are already known to the bot by some other means.
+        users (Sequence[:class:`telegram.SharedUser`]): Information about users shared with the
+            bot. Mutually exclusive with :paramref:`user_ids`.
+
+            .. versionadded:: NEXT.VERSION
+        user_ids (Sequence[:obj:`int`], optional): Identifiers of the shared users. These numbers
+            may have more than 32 significant bits and some programming languages may have
+            difficulty/silent defects in interpreting them. But they have at most 52 significant
+            bits, so 64-bit integers or double-precision float types are safe for storing these
+            identifiers. The bot may not have access to the users and could be unable to use
+            these identifiers, unless the users are already known to the bot by some other means.
+            Mutually exclusive with :paramref:`users`.
+
+            .. versionchanged:: NEXT.VERSION
+                Bot API 7.2 introduced :paramref:`users` replacing this argument. PTB will
+                automatically convert this to that one, but for advanced options, please use
+                :paramref:`users` directly.
+
+            .. deprecated:: NEXT.VERSION
+                In future versions, this argument will become keyword only.
 
     Attributes:
         request_id (:obj:`int`): Identifier of the request.
-        user_ids (Tuple[:obj:`int`]): Identifiers of the shared users. These numbers may have
-            more than 32 significant bits and some programming languages may have difficulty/silent
-            defects in interpreting them. But they have at most 52 significant bits, so 64-bit
-            integers or double-precision float types are safe for storing these identifiers. The
-            bot may not have access to the users and could be unable to use these identifiers,
-            unless the users are already known to the bot by some other means.
+        users (Tuple[:class:`telegram.SharedUser`]): Information about users shared with the
+            bot. Mutually exclusive with :attr:`user_ids`.
+
+            .. versionadded:: NEXT.VERSION
     """
 
-    __slots__ = ("request_id", "user_ids")
+    __slots__ = ("request_id", "users")
 
     def __init__(
         self,
         request_id: int,
-        user_ids: Sequence[int],
+        users: Sequence["SharedUser"],
+        user_ids: Optional[Sequence[int]] = None,
```

Comment:
For backward compatibility I would have expected
```suggestion
        user_ids: Optional[Sequence[int]] = None,
        users: Optional[Sequence["SharedUser"]] = None,
```

The docstrings are also a bit out of order IMO:

* the "mutually exclusive" thing should be in the args section not attrs
* the "changed" note mentions some conversion, but there is not conversion taking place in the `__init__`.
* `users` will become a positional argument, `user_ids` will be removed